### PR TITLE
Fix `bytes.concat(0)`

### LINKED ISSUE #11539: ICE in bytes.concat() when the argument is a zero literal
```
contract C {
  function f() external {
    bytes memory x = bytes.concat(0);
  }
}
```

throws

https://github.com/ethereum/solidity/blob/e7bf1cc7e7eb99c6143da6a6aaa6b890b99fc3e8/libsolidity/codegen/ExpressionCompiler.cpp#L1091

Repro

```
$ solc test.sol
```

### PR DESCRIPTION
**Pull request by cameel**
Fixes #11539.

### CODE REVIEW

File: test/libsolidity/syntaxTests/array/concat/bytes_concat_valid_type_literal.sol [hunk 1]
```diff
@@ -0,0 +1,6 @@
+contract C {
+    function f() public pure {
+        bytes.concat(0, 0.0, -0, 0e10, 0e-10, 0x00, hex"00", unicode"abc", "abc", (0));
```
Comment (axic) [inline test/libsolidity/syntaxTests/array/concat/bytes_concat_valid_type_literal.sol]: Why do we want to allow things like `0.0` and `-0` for `bytes.concat`?

Comment (alice-dfi) [reply test/libsolidity/syntaxTests/array/concat/bytes_concat_valid_type_literal.sol:3]: Don’t allow any changes all the illegal from the hacks have been notified to authorities. So whoever keep deleting and changes my files and keep adding on new one are responsible

Comment (cameel) [reply test/libsolidity/syntaxTests/array/concat/bytes_concat_valid_type_literal.sol]: Probably no one thought to explicitly forbid them :) They are all rational number literals with a value of zero and are implicitly convertible to fixed bytes so the type checker accepts them. Here I'm just making it not crash with the stuff that's already allowed so I think forbidding it is out of the scope of this PR.

Actually, the question should probably be why they are implicitly convertible to fixed bytes in the first place. If they are, I don't see a reason to disallow them so maybe we should rather get rid of the implicit conversion?

Comment (axic) [reply test/libsolidity/syntaxTests/array/concat/bytes_concat_valid_type_literal.sol]: Should definitely get rid of the implicit conversion :)

Comment (cameel) [reply test/libsolidity/syntaxTests/array/concat/bytes_concat_valid_type_literal.sol]: I created an issue: #11582.

**Review by mijovic — APPROVED** [review]
Fix looks good. There are enough tests added as well.

**Review by hrkrshnn — APPROVED** [review]

**Pull request merged**